src/network_subscription.rs

```diff
@@ -0,0 +1,148 @@
+                                    hardware_event_receiver,
+                                    hardware_event_sender,
+                                    connection,
+                                );
+                            }
+                            Err(e) => {
+                                eprintln!("Error connecting to piglet: {e}");
+                            }
+                        }
+                    }
+
+                    NetworkState::Connected(
+                        hardware_event_receiver,
+                        _hardware_event_sender,
+                        connection,
+                    ) => {
+                        let hardware_event = hardware_event_receiver.select_next_some().await;
+                        let mut config_sender = connection.open_uni().await.unwrap();
+
+                        let message = serde_json::to_string(&hardware_event).unwrap();
+                        config_sender.write_all(message.as_bytes()).await.unwrap();
+                        config_sender.finish().await.unwrap();
+                    }
+                }
+            }
+        },
+    )
+}
+
+#[derive(Debug)]
+struct Piglet {
+    /// The id of the remote node.
+    node_id: NodeId,
+    /// The list of direct UDP addresses for the remote node.
+    addrs: Vec<SocketAddr>,
+}
+
+async fn connect() -> anyhow::Result<(HardwareDescription, Connection)> {
+    let args = Piglet {
+        node_id: NodeId::from_str("hyhwn3lk45d76uv6dyoaoeya6ido26tubski6rfum6rjpt2ees6q").unwrap(),
+        addrs: vec![
+            "79.154.163.213:49882".parse().unwrap(),
+            "192.168.1.77:49882".parse().unwrap(),
+        ],
+    };
+
+    let secret_key = SecretKey::generate();
+
+    // Build a `Endpoint`, which uses PublicKeys as node identifiers
+    let endpoint = Endpoint::builder()
+        // The secret key is used to authenticate with other nodes.
+        .secret_key(secret_key)
+        // Set the ALPN protocols this endpoint will accept on incoming connections
+        .alpns(vec![PIGLET_ALPN.to_vec()])
+        // `RelayMode::Default` means that we will use the default relay servers to holepunch and relay.
+        .relay_mode(RelayMode::Default)
+        // You can choose a port to bind to, but passing in `0` will bind the socket to a random available port
+        .bind(0)
+        .await?;
+
+    let me = endpoint.node_id();
+    println!("node id: {me}");
+    println!("node listening addresses:");
+    for local_endpoint in endpoint
+        .direct_addresses()
+        .next()
+        .await
+        .context("no endpoints")?
+    {
+        println!("\t{}", local_endpoint.addr)
+    }
+
+    let relay_url = endpoint.home_relay().ok_or(io::Error::new(
+        io::ErrorKind::Other,
+        "Could not get home relay",
+    ))?;
+
+    // Build a `NodeAddr` from the node_id, relay url, and UDP addresses.
+    let addr = NodeAddr::from_parts(args.node_id, Some(relay_url), args.addrs);
```

Comment:
If you want to simplify this, you can use [`Endpoint::node_addr`](https://docs.rs/iroh-net/latest/iroh_net/endpoint/struct.Endpoint.html#method.node_addr) here and skip the manual extraction of the parts above.